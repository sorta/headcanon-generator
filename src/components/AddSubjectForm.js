import React, { Component } from 'react';
import PropTypes from 'prop-types';

class AddSubjectForm extends Component {
  static propTypes = {
    addSubject: PropTypes.func,
    selectedFandomKey: PropTypes.string,
  };
  static defaultProps = {
    addSubject: () => {},
    selectedFandomKey: '',
  };

  nameRef = React.createRef();

  createSubject = (ev) => {
    ev.preventDefault();
    const subject = {
      name: this.nameRef.current.value,
      fandomKey: this.props.selectedFandomKey,
    };
    this.props.addSubject(this.props.selectedFandomKey, subject);
    ev.currentTarget.reset();
  }

  render() {
    return (
      <form className="AddSubjectForm form-add" onSubmit={this.createSubject}>
        <input name="name" ref={this.nameRef} type="text" placeholder="New Subject Name" required />
        <button type="submit">+</button>
      </form>
    );
  }
}

export default AddSubjectForm;
