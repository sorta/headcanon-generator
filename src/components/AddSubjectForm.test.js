import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

import AddSubjectForm from './AddSubjectForm';

const props = {
  addSubject: jest.fn(),
  isManaging: () => false,
  selectedFandomKey: '',
};

// Tests for regular mode
describe("AddSubjectForm", () => {
  const component = shallow(<AddSubjectForm { ...props } />);

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <AddSubjectForm { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should be null when not managing', () => {
    expect(component.getElement()).toBe(null);
  });
});


// Tests for managing mode
props.isManaging = () => true;
describe("AddSubjectForm in managing mode", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <AddSubjectForm { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have a name input field', () => {
    const component = shallow(<AddSubjectForm { ...props } />);
    expect(component.find('input[name="name"]').exists()).toBe(true);
  });

  it('should have a submit button', () => {
    const component = shallow(<AddSubjectForm { ...props } />);
    expect(component.find('button[type="submit"]').exists()).toBe(true);
  });

  it('form submit should call addSubject', () => {
    const component = mount(<AddSubjectForm { ...props } />);
    component.find('input[name="name"]').instance().value = 'TESTING';
    component.find('button[type="submit"]').simulate('submit');
    expect(props.addSubject.mock.calls.length).toEqual(1);
  });
});
