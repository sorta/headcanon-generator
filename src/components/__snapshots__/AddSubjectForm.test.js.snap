// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddSubjectForm in managing mode renders and matches our snapshot 1`] = `
<form
  className="AddSubjectForm form-add control-row"
  onSubmit={[Function]}
>
  <input
    className="control-stretch"
    name="name"
    placeholder="New Subject Name"
    required={true}
    type="text"
  />
  <button
    type="submit"
  >
    +
  </button>
</form>
`;

exports[`AddSubjectForm renders and matches our snapshot 1`] = `
<form
  className="AddSubjectForm form-add control-row"
  onSubmit={[Function]}
>
  <input
    className="control-stretch"
    name="name"
    placeholder="New Subject Name"
    required={true}
    type="text"
  />
  <button
    type="submit"
  >
    +
  </button>
</form>
`;
